adding gitlab-api required dependencies

dcendents opened the pull request:
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

Adding `jersey-media-json-jackson` dependency required by the `gitlab-api` plugin
https://github.com/jenkinsci/gitlab-api-plugin/pull/7


Also adding a dependency on `jackson2-api` plugin, to prevent a circular dependency error, `jackson2-api` plugin needs to be updated to remove its dependency on `jersey2-api`. `gitlab-plugin` needs to use `jersey-media-json-jackson` instead of `jackson-jaxrs-json-provider`.

A release of the 3 plugins with the following PRs will need to be coordinated:

- https://github.com/jenkinsci/jersey2-api-plugin/pull/9
- https://github.com/jenkinsci/jackson2-api-plugin/pull/122
- https://github.com/jenkinsci/gitlab-plugin/pull/1255

basil reviewed (commented):
Thanks for the PR! Have you been able to confirm that the incremental build from this PR, in concert with the other PRs referenced, resolves the issue?

dcendents commented:
@basil It's hard to say for sure because I was never able to replicate it on demand on my test server (the `java.lang.ClassCastException: Cannot cast org.glassfish.jersey.inject.hk2.Hk2InjectionManagerFactory to org.glassfish.jersey.internal.inject.InjectionManagerFactory`)

If I updated the `gitlab-plugin` (1.5.25 or newer) and had `jersey2-api` installed, the error would eventually come up after some time (normally within a day on our production server).

With this update I tested the normal behaviors of `gitlab-branch-source` plugin
* Scan Project
* Trigger from gitlab webhook
    * branch
    * merge request
* Make sure the gitlab-ci pipeline status is updated correctly


Now since both `gitlab-api` and `jersey2-api` plugins had the `jersey-hk2` lib, I believe the `ClassCastException` above was caused because it is trying to cast a class into its interface but the class and the interface were loaded with 2 different classloaders. My bet is that now by eliminating this second set of jersey libs (from our jenkins server at least), this error should not happen anymore. So far it hasn't but I've installed it on our production server only a few hours ago.

I've seen many people report they had the same problem when the `artifactory` plugin was installed, but we don't use it.

mymarche commented:
I installed this version on my Jenkins, and this works as expected.
@dcendents, Thank you!

basil reviewed (changes requested):
This bundles Jackson JARs, which is not correct because there is a separate Jackson 2 API plugin. Please test #13 instead.

basil commented:
#13 won't work, either. You need to create a third plugin to break the circular dependency between the Jersey 2 API and Jackson 2 API plugins.

dcendents commented:
@basil actually I think it would be preferable to break the dependency jackson2-api -> jersey2-api.

jackson2-api includes the following dependencies:

```xml
    <dependency>
      <groupId>com.fasterxml.jackson.jaxrs</groupId>
      <artifactId>jackson-jaxrs-json-provider</artifactId>
      <version>${revision}</version>
    </dependency>
    <dependency> <!-- needed for jackson-jaxrs-json-provider -->
        <groupId>io.jenkins.plugins</groupId>
        <artifactId>jersey2-api</artifactId>
        <version>2.35-3</version>
        <optional>true</optional>
    </dependency>
```

`jackson-jaxrs-json-provider` is the old dependency that has been superseeded by `jersey-media-json-jackson`.
Jersey need to know about jackson but jackson does not need to know about jersey.

I don't know why both of them decided to provide their own implementation of `JacksonJaxbJsonProvider`, but the correct way of configuring `Jackson` with `Jersey` is with the `JacksonFeature` class which only exists in `jersey-media-json-jackson`.
Actually this is the same change I did in my gitlab4j-api PR https://github.com/gitlab4j/gitlab4j-api/pull/795

Without that jersey will use jsonb (eclipse yasson).

What do you think?
(I suppose if we remove a jar from a plugin we can release it with an incompatible warning?)

basil commented:
`jackson-jaxrs-json-provider` was added to `jackson2-api` for `gitlab-plugin`. So if we want to remove `jackson2-api`'s dependency on `jersey2-api`, then we need to:

- Open a draft PR against `jackson2-api` to remove the dependency on `jackson-jaxrs-json-provider` and `jersey2-api`.
- Open a draft PR against `gitlab-plugin` consuming the incremental from `jackson2-api`. Test the resulting build of `gitlab-plugin` to make sure it still works without `jackson-jaxrs-json-provider`. If it doesn't work, fix it (probably by migrating to `jersey-media-json-jackson`?).
- Update this PR with the Jackson portions of #13, consuming the incremental of `jackson2-api`.

If draft PRs of `jackson2-api`, `jersey2-api`, and `gitlab-plugin` can all pass CI and manual testing, then we can talk about a merge/release plan. Also note that I am not a maintainer of `jackson2-api`.

basil commented:
I added some cleanup in commit 31f10d84d0ff08d9c7eea7c9222f1a2fd7af0a6a. @dcendents Can you please move `JacksonJsonProviderTest` from `jackson2-api` to `jersey2-api`, adapting it as necessary?

basil commented on pom.xml:
```diff
@@ -101,6 +130,12 @@
+    <dependency>
+      <groupId>org.jenkins-ci.plugins</groupId>
+      <artifactId>jackson2-api</artifactId>
+      <version>2.13.1-257.vb_254f61c3a_2b_</version> <!-- TODO https://github.com/jenkinsci/jackson2-api-plugin/pull/122 -->
```
jenkinsci/jackson2-api-plugin#122

dcendents commented:
> I added some cleanup in commit [31f10d8](https://github.com/jenkinsci/jersey2-api-plugin/commit/31f10d84d0ff08d9c7eea7c9222f1a2fd7af0a6a). @dcendents Can you please move `JacksonJsonProviderTest` from `jackson2-api` to `jersey2-api`, adapting it as necessary?

I've moved the test and adapted it to load a `JacksonFeature` (which itself registers a `JacksonJaxbJsonProvider`, an extension of `JacksonJsonProvider`).

We could also test `JacksonJsonProvider` and `JacksonJaxbJsonProvider` but since they are in a package named internal (`org.glassfish.jersey.jackson.internal.jackson.jaxrs.json`) I think it is preferable we stick with `JacksonFeature` only.

dcendents commented:
I just got the following error in Jenkins:

```
Exception caught: class org.glassfish.jersey.jaxb.internal.JaxbAutoDiscoverable cannot be cast to class org.glassfish.jersey.internal.spi.AutoDiscoverable (org.glassfish.jersey.jaxb.internal.JaxbAutoDiscoverable is in unnamed module of loader jenkins.util.AntWithFindResourceClassLoader @6853b21a; org.glassfish.jersey.internal.spi.AutoDiscoverable is in unnamed module of loader jenkins.util.AntWithFindResourceClassLoader @5764ba86)
```

Can this be related to `javax-activation-api` plugin?

basil commented:
> Can this be related to `javax-activation-api` plugin?

I don't see how it could be given that neither the error message nor any classes referenced in the error message contain references to `javax.activation` or `jakarta.activation`. Seems like a separate problem best discussed in another PR. See also jenkinsci/jaxb-plugin#15, which appears to be stalled due to an unresponsive maintainer.

basil commented:
> I just got the following error in Jenkins:
> 
> Exception caught: class org.glassfish.jersey.jaxb.internal.JaxbAutoDiscoverable cannot be cast to class org.glassfish.jersey.internal.spi.AutoDiscoverable (org.glassfish.jersey.jaxb.internal.JaxbAutoDiscoverable is in unnamed module of loader jenkins.util.AntWithFindResourceClassLoader @6853b21a; org.glassfish.jersey.internal.spi.AutoDiscoverable is in unnamed module of loader jenkins.util.AntWithFindResourceClassLoader @5764ba86)

Try looking through the extracted plugin `.jpi` files in `${JENKINS_HOME}/plugins}` to see which plugins contain a JAR with `JaxbAutoDiscoverable.class`.

basil commented:
CI builds of all three PRs are passing:

- https://ci.jenkins.io/job/Plugins/job/jackson2-api-plugin/job/PR-122/
- https://ci.jenkins.io/job/Plugins/job/jersey2-api-plugin/job/PR-9/
- https://ci.jenkins.io/job/Plugins/job/gitlab-plugin/job/PR-1255/

Next step, I think, is to do some real-world testing of the latest incrementals.

basil commented:
> I've moved the test and adapted it to load a `JacksonFeature` (which itself registers a `JacksonJaxbJsonProvider`, an extension of `JacksonJsonProvider`).

Sure, but the test wasn't invoking this registration because we weren't calling `JacksonFeature#configure` from the test. I've improved the test in commit c420823129f669c10e1a6f18d24d97a80d69e386 and confirmed that the test now successfully creates an instance of `org.glassfish.jersey.jackson.internal.DefaultJacksonJaxbJsonProvider`.

dcendents commented:
> Try looking through the extracted plugin `.jpi` files in `${JENKINS_HOME}/plugins}` to see which plugins contain a JAR with `JaxbAutoDiscoverable.class`.

I can confirm the only other copy of `JaxbAutoDiscoverable.class` was in `jira/WEB-INF/lib/jersey-media-jaxb2.35.jar`

After disabling the `jira` plugin, I re-ran the same tests (create a new branch in gitlab and a merge request, make sure jenkins starts a build and notifies gitlab-ci of the pipeline stage) and I didn't get an error this time.

I've installed it on my production server and will monitor it closely for the next few days.

basil commented:
> I can confirm the only other copy of `JaxbAutoDiscoverable.class` was in `jira/WEB-INF/lib/jersey-media-jaxb2.35.jar`

Yeah, looks like `jira` should be updated to stop bundling Jersey and instead depend on the `jersey2-api` plugin.

basil commented:
Released in [2.35-6](https://github.com/jenkinsci/jersey2-api-plugin/releases/tag/jersey2-api-2.35-6).

basil commented:
Thanks for the PR, @dcendents! Hopefully this won't cause too many problems now that it's released. :sweat_smile:

> I can confirm the only other copy of `JaxbAutoDiscoverable.class` was in `jira/WEB-INF/lib/jersey-media-jaxb2.35.jar`
> 
> After disabling the `jira` plugin, I re-ran the same tests (create a new branch in gitlab and a merge request, make sure jenkins starts a build and notifies gitlab-ci of the pipeline stage) and I didn't get an error this time.

If you're interested in working on the Jira plugin next, I opened a draft to add `jersey-media-json-jettison` to `jersey2-api` in https://github.com/jenkinsci/jersey2-api-plugin/pull/19, and it passes the `gitlab-plugin` test suite in https://github.com/jenkinsci/gitlab-plugin/pull/1276. I think all that would be needed to update `jira` is something like

```xml
diff --git a/pom.xml b/pom.xml
index a6c7fe6..2f4c18e 100644
--- a/pom.xml
+++ b/pom.xml
@@ -174,6 +174,24 @@
           <groupId>org.codehaus.jackson</groupId>
           <artifactId>*</artifactId>
         </exclusion>
+
+        <!-- Provided by Jersey 2 API plugin -->
+        <exclusion>
+          <groupId>org.glassfish.jersey.core</groupId>
+          <artifactId>jersey-client</artifactId>
+        </exclusion>
+        <exclusion>
+          <groupId>org.glassfish.jersey.core</groupId>
+          <artifactId>jersey-common</artifactId>
+        </exclusion>
+        <exclusion>
+          <groupId>org.glassfish.jersey.media</groupId>
+          <artifactId>jersey-media-jaxb</artifactId>
+        </exclusion>
+        <exclusion>
+          <groupId>org.glassfish.jersey.media</groupId>
+          <artifactId>jersey-media-json-jettison</artifactId>
+        </exclusion>
       </exclusions>
     </dependency>
     <dependency>
@@ -224,9 +242,18 @@
       <artifactId>mailer</artifactId>
     </dependency>
 
+    <dependency>
+      <groupId>io.jenkins.plugins</groupId>
+      <artifactId>jersey2-api</artifactId>
+      <!-- TODO remove the explicit version number once this has made it into the plugin BOM -->
+      <version>2.35-7-rc39.1c80918c8b_a_6</version> <!-- TODO https://github.com/jenkinsci/jersey2-api-plugin/pull/19 -->
+    </dependency>
+
     <dependency>
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>jackson2-api</artifactId>
+      <!-- TODO remove the explicit version number once this has made it into the plugin BOM -->
+      <version>2.13.2.20220328-273.v11d70a_b_a_1a_52</version>
```

although I have no way of testing it and I lack the time to propose a PR to the Jira plugin.

dcendents commented:
@basil thanks for merging them

I don't really use the JIRA plugin either (although we have JIRA), we only got it in jenkins because it gets installed automatically by Blue Ocean.

Next I have to fix gitlab-api and gitlab branch source plugins. Neither seem to be actively maintained though.

basil commented:
Nice, let me know if I can help in any way.

jetersen commented:
@dcendents if you need, I can review and release gitlab-api and gitlab-branch-source :)
